exporters/pelorus/deserialization/__init__.py

```diff
@@ -79,54 +83,119 @@
     MissingFieldError,
     TypeCheckError,
 )
-from pelorus.utils import BadAttributePathError, get_nested
+from pelorus.utils import BadAttributePathError, format_path, get_nested, split_path
 from pelorus.utils._attrs_compat import NOTHING
 
 # metadata
 _NESTED_PATH_KEY = "__pelorus_structure_nested_path"
+_RETAIN_SOURCE_KEY = "__pelorus_structure_retain_source"
 
 T = TypeVar("T")
 
 PRIMITIVE_TYPES: tuple[type, ...] = (int, float, str, bool)
 
 
-def nested(path: str) -> dict[str, Any]:
-    "Put the nested data path in a metadata dict."
-    return {_NESTED_PATH_KEY: path}
+def nested(path: Union[str, Sequence[str]]) -> dict[str, Sequence[str]]:
+    """
+    Put the nested data path in a metadata dict.
+    Like get_nested, use a list if a path segment has a dot in it.
+    """
+    return {_NESTED_PATH_KEY: split_path(path)}
 
 
-def _is_attrs_class(cls: type) -> Optional[type["attr.AttrsInstance"]]:
+def retain_source(retain: bool = True) -> dict[str, bool]:
+    """
+    Retain the source of the deserialization in this field.
+    """
+    return {_RETAIN_SOURCE_KEY: retain}
+
+
+# python has some magic to make classes work with `issubclass` and instances work with `isinstance`,
+# even if they don't actually inherit from them. Some so-called "abstract base classes" just check
+# for the presence of certain methods. Cool!
+# However, some of them don't support that, for reasons outlined in their docs (see the first section):
+# https://docs.python.org/3.11/library/collections.abc.html
+
+# thus, we need to make our own checks. These have the same downsides as in the docs,
+# but they're "good enough". "Practicality beats purity".
+
+# we need this for mappings specifically, because a kubernetes.dynamic.resource.ResourceField has
+# __getitem__, which is actually all we care about.
```

Comment:
https://mypy.readthedocs.io/en/stable/cheat_sheet_py3.html#standard-duck-types type them as sequence would fix the problem? (kubernetes.dynamic.resource.ResourceField should have len as well?)

Reply:
Discussed in meeting :+1: